### PR TITLE
Feat(web-react): Introduce `Breadcrumbs` component (refs #DS-302)

### PR DESCRIPTION
**Pull request by vmilersky**

### CODE REVIEW

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  return (
+    <ElementTag
+      {...otherProps}
+      {...styleProps}
+      className={classNames(classProps, styleProps.className)}
+      aria-label="Breadcrumb"
+    >
+      <ol>
+        {restProps.items.map((item, index) => (
+          <React.Fragment key={`BreadcrumbsItem_${item.url}`}>
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:25]: Are you sure that `item.url` is a great candidate for React `key` prop? Especially with all those weird strings `//:?&@`... I would suggest using `item.title` as a better candidate. What do you think?

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  return (
+    <ElementTag
+      {...otherProps}
+      {...styleProps}
+      className={classNames(classProps, styleProps.className)}
+      aria-label="Breadcrumb"
+    >
+      <ol>
+        {restProps.items.map((item, index) => (
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:24]: I think it would be better to destructure `restProps` and use only `items`.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+          <React.Fragment key={`BreadcrumbsItem_${item.url}`}>
+            {index === 0 && (
+              <li className="d-tablet-none">
+                <a href={item.url}>Back</a>
+              </li>
+            )}
+            <li className="d-none d-tablet-flex">
+              <a href={item.url} aria-current={index === restProps.items.length - 1 ? 'page' : undefined}>
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:32]: I am thinking about making this condition more readable and defining an arrow function for this. But I am quite not sure if it is necessary. What is your opinion?

```suggestion
              <a href={item.url} aria-current={isCurrentPage(index, itemsCount) ? 'page' : undefined}>
```

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import classNames from 'classnames';
+import { useBreadcrumbsStyleProps } from './useBreadcrumbsStyleProps';
+import { SpiritBreadcrumbsProps } from '../../types';
+import { useStyleProps } from '../../hooks/styleProps';
+
+const defaultProps = {
+  items: [],
+};
+
+export const Breadcrumbs = <T extends ElementType = 'nav'>(props: SpiritBreadcrumbsProps<T>): JSX.Element => {
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:11]: In https://github.com/lmc-eu/spirit-design-system/pull/431/files#diff-208644f2adb988532a0e8f1ac1350099cff104a6d3cfce2a8a7f407cb2fa5641 you are also defining `BreadcrumbItem` and `BreadcrumbLink`. Why you do not do it here in the same way?

**Review by literat — CHANGES_REQUESTED** [review]

**Comment by literat** [comment]
Super excited to have a new component here 👍 I've also written down some suggestions and questions to answer.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      {...styleProps}
+      className={classNames(classProps, styleProps.className)}
+      aria-label="Breadcrumb"
+    >
+      <ol>
+        {restProps.items.map((item, index) => (
+          <React.Fragment key={`BreadcrumbsItem_${item.url}`}>
+            {index === 0 && (
+              <li className="d-tablet-none">
+                <a href={item.url}>Back</a>
```
[the remark this answers is not in the document]
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx]: That's true, I moved it to props 👍

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 1 above)
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx]: Well I tried to use `index` since there shouldn't be any re-rendering issues within this component but eslint told me not to do that. OK, I changed `item.url` to `item.title`. I think that breadcrumbs should not contain any item that matches url nor title anyway.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 3 above)
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:33]: Since you think this is not a very nice markdown for you, I suggest to make a function `isLast` because that's what it's supposed to say, not that the page is a current page. There's no url path and breadcrumb url item comparison. Do you agree?

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 4 above)
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:12]: I wrote the React component first but you started reading first the Twig component :-) Twig version's structure is similar to the Tabs component, that's what I was getting inspiration from. The React component is much simpler than the Twig one so I didn't want to tear it apart and just add extra redundant code just to keep the same structure. I hope I understood your comment well.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 2 above)
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:25]: I tried to destructure it but `useBreadcrumbsStyleProps` expects the `SpiritBreadcrumbsProps` (same as other components do).

I think I need some help here, please.

Screenshot 2022-08-22 at 13 59 35

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 4 above)
Comment (literat) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:12]: I understand your intention but I think that you are missing a few points.

- First of all, you are reusing code and defining anchor and list two times.
- Secondly, switching the class names should be based on some prop, because it is a state of a component, in this case `<BreadcrumbItem isMobileBack={true} />` (or something similar).
- Thirdly, I think that mobile Back breadcrumb should target not root item but last - 1. User want to go back one page, not to the root.

I think that we want to have a structure like this and let the developer using this component decide how to generate breadcrumbs in an app's code.

```jsx
<Breadcrumbs>
  <BreadcrumbItem url="#" isMobile={true}>Back</BreadcrumbItem>
  <BreadcrumbItem url="#">Root</BreadcrumbItem>
  <BreadcrumbItem url="#">Category</BreadcrumbItem>
</Breadcrumbs>
```

or let the developer do

```jsx
<Breadcrumbs>
  {items.map(item => <BreadcrumbItem url={item.url}>{item.title}</BreadcrumbItem> )}
</Breadcrumbs>
```

Please, read also comment from @adamkudrna - https://github.com/lmc-eu/spirit-design-system/issues/365

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 2 above)
Comment (literat) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:25]: I this case the `useBreadcrumbsStyleProps` should expect only `BreadcrumbsProps`, because you destructure them in the component `const { elementType: ElementTag = 'nav', items, goBackItemTranslation, ...restProps } = props;`. You do not want to let the `items` and `goBackItemTranslation` go to `{...otherProps}` because they are not valid HTML attributes.

File: packages/web-react/src/components/Breadcrumbs/useBreadcrumbsStyleProps.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+export interface BreadcrumbsStyles {
+  /** className props */
+  classProps: string;
+  /** props to be passed to the element */
+  props: BreadcrumbsProps;
+}
+
+export function useBreadcrumbsStyleProps<T extends ElementType = 'nav'>(
+  props: SpiritBreadcrumbsProps<T>,
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/useBreadcrumbsStyleProps.tsx:13]:
```suggestion
  props: BreadcrumbsProps,
```

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import { SpiritBreadcrumbsProps } from '../../types';
+import { useStyleProps } from '../../hooks/styleProps';
+
+const defaultProps = {
+  goBackItemTranslation: 'Back',
+  items: [],
+};
+
+export const Breadcrumbs = <T extends ElementType = 'nav'>(props: SpiritBreadcrumbsProps<T>): JSX.Element => {
+  const { elementType: ElementTag = 'nav', ...restProps } = props;
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:13]:
```suggestion
  const { elementType: ElementTag = 'nav', items, goBackItemTranslation, ...restProps } = props;
```

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 4 above)
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:12]: There is another use case that cannot be applied on this solution and it's called "microdata". If a developer will need to make code richer by microdata, he just can't .

After a discussion we agreed that re-done of this young component is necessary.

Component should (as its React brother component) be able to accept array of items that it renders out of the box, just by itself.

Developers have to be able to render completely different content of this component like microdata usage mentioned above.

The component was reworked and is ready for its "another first review". 

Please check it out @literat

File: packages/web-react/src/components/Breadcrumbs/useBreadcrumbsStyleProps.tsx (hunk 6 above)
Comment (vmilersky) [reply packages/web-react/src/components/Breadcrumbs/useBreadcrumbsStyleProps.tsx]: Yes, that's what I meant in comment above.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 4 above)
Comment (literat) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:12]: I think that supporting microdata now is out of scope.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      {...styleProps}
+      className={classNames(classProps, styleProps.className)}
+      aria-label="Breadcrumb"
+    >
+      <ol>
+        {items.map((item, index) => (
+          <React.Fragment key={`BreadcrumbsItem_${item.title}`}>
+            {index === 0 && (
+              <li className="d-tablet-none">
+                <a href={item.url}>{goBackItemTranslation}</a>
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:33]: Should we invent a better name for this prop, please? For example `goBackLabel` or `goBackTitle`? I think `goBackItemTranslation` is too long and there is no such thing as translating string so it can be misunderstood.

File: packages/web-react/src/components/Breadcrumbs/README.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+```jsx
+<Breadcrumbs items={items} />
+```
+
+## Available props
+
+| Name    | Type                           | Description           |
+| ------- | ------------------------------ | --------------------- |
+| `items` | `{ title: 'Title', url: '#' }` | Navigation menu items |
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/README.md:36]: There is a missing documented prop for a go-back item.

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 4 above)
Comment (vmilerskyV1) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:12]: As we agreed the component lets developer to choose whether to pass breadcrumb items array or to define custom ordered list's content.

Thank you, good point that the mobile devices back link href value was wrong. It's been fixed.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 8 above)
Comment (vmilerskyV1) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx]: `goBackTitle` was a bit unclear or on the other hand much simple for me, but if you're OK with that, I'm as well. 
It's been changed now.

File: packages/web-react/src/components/Breadcrumbs/README.md (hunk 9 above)
Comment (vmilerskyV1) [reply packages/web-react/src/components/Breadcrumbs/README.md]: README was updated.

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import React, { ElementType } from 'react';
+import classNames from 'classnames';
+import { useBreadcrumbsStyleProps } from './useBreadcrumbsStyleProps';
+import { SpiritBreadcrumbsProps } from '../../types';
+import { useStyleProps } from '../../hooks/styleProps';
+
+const defaultProps = {
+  goBackTitle: 'Back',
```
Comment (literat) [inline packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx:8]: I think that this default should also be empty. We do not want to handle any default strings or their translations. `goBackTitle` should be applied only when there is an`items` prop used. When using `children`, there should be no default or it should be empty.

Typescript cannot work simply with mutually exclusive types, which `children` and `items` in this case are. So we must let the developer decide which one he/she wants to use.

Please set the default to an empty string.

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 10 above)
Comment (vmilerskyV1) [reply packages/web-react/src/components/Breadcrumbs/Breadcrumbs.tsx]: Default value is empty string now.

File: packages/web/src/scss/components/Breadcrumbs/_theme.scss [hunk 11]
```diff
@@ -12,7 +12,7 @@ $link-color-default: tokens.$action-link-primary-default;
 $link-color-active: tokens.$action-link-primary-active;
 $link-color-hover: tokens.$action-link-primary-hover;
 
-$current-link-color: tokens.$text-secondary-default;
+$current-link-color: tokens.$action-link-secondary-default;
```
Comment (literat) [inline packages/web/src/scss/components/Breadcrumbs/_theme.scss:15]: IMHO, this should be in a separate commit or PR.

**Review by literat — APPROVED** [review]

File: packages/web/src/scss/components/Breadcrumbs/_theme.scss (hunk 11 above)
Comment (vmilersky) [reply packages/web/src/scss/components/Breadcrumbs/_theme.scss]: OK, I moved it here https://github.com/lmc-eu/spirit-design-system/pull/452
